### PR TITLE
Topology: Keyword Detection specific changes

### PR DESCRIPTION
**Pull request by keyonjie**
This should be applied after https://github.com/thesofproject/sof/pull/1079 is merged.

### CODE REVIEW

File: tools/topology/m4/eq_fir.m4 [hunk 1]
```diff
@@ -28,9 +28,9 @@ define(`W_EQ_FIR',
 `	tuples "'N_EQ_FIR($1)`_tuples_str"'
 `}'
 `SectionVendorTuples."'N_EQ_FIR($1)`_tuples_str_type" {'
-`	tokens "sof_effect_tokens"'
+`	tokens "sof_process_tokens"'
```
Comment (tlauda) [inline tools/topology/m4/eq_fir.m4]: Already covered by #1079.

File: tools/topology/m4/pipeline.m4 [hunk 2]
```diff
@@ -91,7 +91,7 @@ define(`PIPELINE_PCM_DAI_ADD',
 
 dnl PIPELINE_ADD(pipeline,
 dnl     pipe id, max channels, format,
-dnl     frames, deadline, priority, core)
+dnl     frames, deadline, priority, core, sched_comp)
```
Comment (tlauda) [inline tools/topology/m4/pipeline.m4]: Already covered by #1075.

File: tools/topology/sof-apl-keyword-detect.m4 [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+dnl PIPELINE_PCM_ADD(pipeline,
+dnl     pipe id, pcm, max channels, format,
+dnl     frames, deadline, priority, core)
+
+# Passthrough capture pipeline 1 on PCM 0 using max 2 channels.
+# Schedule 16 frames per 1000us deadline on core 0 with priority 0
+PIPELINE_PCM_ADD(sof/pipe-kfbm-capture.m4,
+	1, 0, 2, s32le,
+	16, 1000, 0, 0)
```
Comment (tlauda) [inline tools/topology/sof-apl-keyword-detect.m4]: 320 frames for 20 ms periods.

File: tools/topology/sof-apl-keyword-detect.m4 [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+dnl PCM_CAPTURE_ADD(name, pipeline, capture)
+PCM_CAPTURE_ADD(DMIC01, 1, PIPELINE_PCM_1)
+
+# keyword detector pipe
+dnl PIPELINE_ADD(pipeline,
+dnl     pipe id, max channels, format,
+dnl     frames, deadline, priority, core)
+PIPELINE_ADD(sof/pipe-detect.m4, 2, 2, s16le, 16, 1000, 0, 0, PIPELINE_SCHED_COMP_1)
```
Comment (tlauda) [inline tools/topology/sof-apl-keyword-detect.m4]: 320 frames for 20 ms periods.

Comment (ranj063) [reply tools/topology/sof-apl-keyword-detect.m4]: @keyonjie the frame size and deadline should be defined in the top level m4 and be common for all pipelines in the topology.

Comment (tlauda) [reply tools/topology/sof-apl-keyword-detect.m4]: @ranj063 @keyonjie Only for keyword detector related pipelines. All other capture and render streams should operate on normal 1 ms frames.

Comment (ranj063) [reply tools/topology/sof-apl-keyword-detect.m4]: @tlauda I'd assume that both the pipelines in sof-apl-keyword-detect.m4 would use the same frame size and deadline. Or do you mean to say only the slector->detect pipeline should use this?

**Comment by ranj063** [comment]
@keyonjie lets break this PR into individual ones say 1 for KPB tplg support, 1 for detect tplg support and another one for keyword-detect tplg addition.

File: tools/topology/m4/detect.m4 [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+`	tuples."word" {'
+`		SOF_TKN_COMP_PERIOD_SINK_COUNT'		STR($3)
+`		SOF_TKN_COMP_PERIOD_SOURCE_COUNT'	STR($4)
+`	}'
+`}'
+`SectionVendorTuples."'N_DETECT($1)`_detect_process_tuples_str" {'
+`	tokens "sof_process_tokens"'
+`	tuples."string" {'
+`		SOF_TKN_PROCESS_TYPE'	STR(concat($5, _DETECT))
+`		SOF_TKN_PROCESS_SUBTYPE	"SIGSINK"'
```
Comment (ranj063) [inline tools/topology/m4/detect.m4]: @keyonjie i discussed with @plbossart about this and we should use the snd_soc_dapm_sink type for the detect component. This way we can avoid the subtype. In the kernel this widget will use the same ipc struct as the process to keep changes minimal. So lets remove this token.

File: tools/topology/m4/detect.m4 [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+`		SOF_TKN_COMP_FORMAT'	STR($2)
+`	}'
+`}'
+`SectionData."'N_DETECT($1)`_data_str" {'
+`	tuples "'N_DETECT($1)`_tuples_str"'
+`	tuples "'N_DETECT($1)`_detect_process_tuples_str"'
+`}'
+`SectionWidget."'N_DETECT($1)`" {'
+`	index "'PIPELINE_ID`"'
+`	type "effect"'
```
Comment (ranj063) [inline tools/topology/m4/detect.m4]: @keyonjie this should be sink

File: tools/topology/m4/detect.m4 [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+`	no_pm "true"'
+`	event_flags	"15"'
```
Comment (ranj063) [inline tools/topology/m4/detect.m4]: @keyonjie can we please add a comment here as to what the 15 and 1 signify. It will be good for future reference.

File: tools/topology/sof/pipe-kfbm-capture.m4 [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+P_GRAPH(pipe-kpbm-capture-PIPELINE_ID, PIPELINE_ID,
+	LIST(`		',
+	`dapm(N_PCMC(PCM_ID), N_BUFFER(1))',
+	`dapm(N_BUFFER(1), N_KPBM(0, PIPELINE_ID))',
+	`dapm(N_KPBM(0, PIPELINE_ID), N_BUFFER(0))'))
+
+#
+# Pipeline Source and Sinks
+# TODO add others for MUX (list ??)
```
Comment (ranj063) [inline tools/topology/sof/pipe-kfbm-capture.m4]: @keyonjie what list is this?

File: tools/topology/m4/ch_sel.m4 [hunk 9]
```diff
@@ -0,0 +1,46 @@
+divert(-1)
+
+dnl Define macro for channel selector widget
+
+dnl Selector name)
+define(`N_SELECTOR', `SELECTOR'PIPELINE_ID`.'$1)
+
+dnl W_SELECTOR(name, format, periods_sink, periods_source, in_chan_cnt, out_chan_cnt, sel_chan_index)
```
Comment (ranj063) [inline tools/topology/m4/ch_sel.m4]: @keyonjie we should probably add the tokens tuples section back too

File: tools/topology/sof/pipe-detect.m4 [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+	 PLATFORM_HOST_MEM_CAP)
+# Capture Buffers
+W_BUFFER(2, COMP_BUFFER_SIZE(2,
+	 COMP_SAMPLE_SIZE(PIPELINE_FORMAT), PIPELINE_CHANNELS, SCHEDULE_FRAMES),
+	 PLATFORM_HOST_MEM_CAP)
+#virtual output widget
+VIRTUAL_WIDGET(DETECT SINK PIPELINE_ID, output, PIPELINE_ID)
+
+#FIXME: platform data in pipeline
+W_PIPELINE(SCHED_COMP, SCHEDULE_DEADLINE, SCHEDULE_PRIORITY, SCHEDULE_FRAMES, SCHEDULE_CORE, 1, pipe_media_schedule_plat)
```
Comment (ranj063) [inline tools/topology/sof/pipe-detect.m4]: @keyonjie unset timer here as it will be dai driven

File: tools/topology/m4/pipeline.m4 (hunk 2 above)
Comment (keyonjie) [reply tools/topology/m4/pipeline.m4]: @ranj063 closed #1075

File: tools/topology/m4/eq_fir.m4 (hunk 1 above)
Comment (keyonjie) [reply tools/topology/m4/eq_fir.m4]: yes, as it has dependency, I apply this on top of #1079 applied, will update it once #1079 merged.

File: tools/topology/sof-apl-keyword-detect.m4 (hunk 3 above)
Comment (keyonjie) [reply tools/topology/sof-apl-keyword-detect.m4]: We haven't verified that if 320 frames(20ms period) work, so I prefer to update it later.

File: tools/topology/sof-apl-keyword-detect.m4 (hunk 4 above)
Comment (keyonjie) [reply tools/topology/sof-apl-keyword-detect.m4]: ditto.

File: tools/topology/sof-apl-keyword-detect.m4 (hunk 3 above)
Comment (plbossart) [reply tools/topology/sof-apl-keyword-detect.m4]: please make that value configurable without delay so we don't have to comb through the topology later...even if you keep the same value for now the ask is that with a single line change all the topology uses the SAME value. Makes sense?

Comment (keyonjie) [reply tools/topology/sof-apl-keyword-detect.m4]: sure, I am waiting for @tlauda's confirmation if we do need 320 frames for the host capture pipeline.

**Comment by keyonjie** [comment]
> @keyonjie lets break this PR into individual ones say 1 for KPB tplg support, 1 for detect tplg support and another one for keyword-detect tplg addition.

Doesn't that really make sense? It is already belong to different commits, they need to work together, in one shot.

File: tools/topology/m4/detect.m4 (hunk 5 above)
Comment (keyonjie) [reply tools/topology/m4/detect.m4]: @ranj063 per your latest comments in kernel PR, we can't use snd_soc_dapm_sink as it is with always on PM, so let me resolve this conversation.

File: tools/topology/m4/detect.m4 (hunk 6 above)
Comment (keyonjie) [reply tools/topology/m4/detect.m4]: I recheck it is effect on your branch, anything I missed?
https://github.com/ranj063/sof/blob/topic/keyword_detect/tools/topology/m4/detect.m4

File: tools/topology/m4/detect.m4 (hunk 7 above)
Comment (keyonjie) [reply tools/topology/m4/detect.m4]: good suggestion, let me add it.

File: tools/topology/sof/pipe-kfbm-capture.m4 (hunk 8 above)
Comment (keyonjie) [reply tools/topology/sof/pipe-kfbm-capture.m4]: no need, let me remove it.

File: tools/topology/sof/pipe-detect.m4 (hunk 10 above)
Comment (keyonjie) [reply tools/topology/sof/pipe-detect.m4]: OK, let's switch it when timer available and needed.

**Comment by keyonjie** [comment]
@ranj063 @tlauda @lgirdwood @plbossart comments addressed and updated.

File: tools/topology/sof-apl-keyword-detect.m4 (hunk 3 above)
Comment (tlauda) [reply tools/topology/sof-apl-keyword-detect.m4]: @keyonjie We need 320 frames for both kwd pipelines. If by host capture you mean the one with dai, kpb and host components, then yes.

File: tools/topology/sof-apl-keyword-detect.m4 (hunk 4 above)
Comment (tlauda) [reply tools/topology/sof-apl-keyword-detect.m4]: @ranj063 @tlauda Yes, for both should be 20 ms.

Comment (keyonjie) [reply tools/topology/sof-apl-keyword-detect.m4]: @tlauda so actually all buffers in this tplg file need to be 20ms.

And with 1 period or 2?

Comment (tlauda) [reply tools/topology/sof-apl-keyword-detect.m4]: @keyonjie 2 unless we confirm that 1 also works.

Comment (keyonjie) [reply tools/topology/sof-apl-keyword-detect.m4]: OK, let me update it like that.

File: tools/topology/m4/detect.m4 (hunk 6 above)
Comment (ranj063) [reply tools/topology/m4/detect.m4]: @keyonjie all good

File: tools/topology/m4/detect.m4 [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+`	event_type	"1"' # 1 for DAPM event for detect component
+`	data ['
+`		"'N_DETECT($1)`_data_w"'
+`		"'N_DETECT($1)`_data_str"'
+`	]'
+`	mixer ['
+		$6
+`	]'
+`	enum ['
+`		$7'
```
Comment (ranj063) [inline tools/topology/m4/detect.m4]: @keyonjie I think we need bytes control too to send the binary data for detection

File: tools/topology/sof/tokens.m4 [hunk 12]
```diff
@@ -88,3 +88,10 @@ SectionVendorTokens."sof_tone_tokens" {
 SectionVendorTokens."sof_effect_tokens" {
 	SOF_TKN_EFFECT_TYPE			"900"
 }
+
+SectionVendorTokens."sof_ch_sel_tokens" {
```
Comment (ranj063) [inline tools/topology/sof/tokens.m4]: @keyonjie we dont need this anymore. lets drop it.

File: tools/topology/sof-apl-keyword-detect.m4 [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+                dapm(PIPELINE_SINK_2, PIPELINE_SOURCE_1)
+        ]
+}
+# Connect pipelines together
+SectionGraph."pipe-sof-apl-keyword-detect" {
+        index "2"
+
+        lines [
+		# keyword detect
+		dapm(PIPELINE_PCM_1, PIPELINE_DETECT_2)
```
Comment (ranj063) [inline tools/topology/sof-apl-keyword-detect.m4]: @keyonjie can we add both these in the same SectionGraph? I added 2 sections because I was trying something silly but I think we can merge it tomake the topology file a bit compact.

File: tools/topology/sof/pipe-dai-capture.m4 [hunk 14]
```diff
@@ -20,3 +20,5 @@ W_PIPELINE(N_DAI_IN, SCHEDULE_DEADLINE, SCHEDULE_PRIORITY, SCHEDULE_FRAMES, SCHE
 
 P_GRAPH(DAI_NAME, PIPELINE_ID,
 	LIST(`		', `dapm(DAI_BUF, N_DAI_IN)'))
+
+indir(`define', concat(`PIPELINE_SCHED_COMP_', PIPELINE_ID), N_DAI_IN)
```
Comment (ranj063) [inline tools/topology/sof/pipe-dai-capture.m4]: @keyonjie this should be a separate commit

File: tools/topology/m4/kpbm.m4 [hunk 15]
```diff
@@ -0,0 +1,49 @@
+divert(-1)
+
+dnl Define macro for Key Phrase Buffer Manager(kpbm) widget
+
+dnl kpbm Name)
+define(`N_KPBM', `KPBM'$2`.'$1)
+
+dnl Pipe Buffer name in pipeline (pipeline, buffer)
+define(`NPIPELINE_MUX', `MUX'$1`.'$2)
```
Comment (ranj063) [inline tools/topology/m4/kpbm.m4]: @keyonjie what is this used for?

Comment (keyonjie) [reply tools/topology/m4/kpbm.m4]: it's sth residual, let me clean it.

File: tools/topology/sof/pipe-dai-capture.m4 (hunk 14 above)
Comment (keyonjie) [reply tools/topology/sof/pipe-dai-capture.m4]: OK, let me add a commit with it as the first one.

File: tools/topology/sof-apl-keyword-detect.m4 (hunk 13 above)
Comment (keyonjie) [reply tools/topology/sof-apl-keyword-detect.m4]: not sure if put them to same index "0" work.

**Comment by keyonjie** [comment]
@juimonen this is updated with integration verification passed. I think it's fine to be merged.

File: tools/topology/m4/detect.m4 (hunk 11 above)
Comment (keyonjie) [reply tools/topology/m4/detect.m4]: let me drop it with FW change and bytes control ready for channel selector and kpbm ready later. @juimonen

File: tools/topology/sof/tokens.m4 (hunk 12 above)
Comment (keyonjie) [reply tools/topology/sof/tokens.m4]: let me drop it with FW change and bytes control ready for channel selector and kpbm ready later. @juimonen

File: tools/topology/m4/detect.m4 [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+`SectionData."'N_DETECT($1)`_data_str" {'
+`	tuples "'N_DETECT($1)`_tuples_str"'
+`	tuples "'N_DETECT($1)`_detect_process_tuples_str"'
+`}'
+`SectionWidget."'N_DETECT($1)`" {'
+`	index "'PIPELINE_ID`"'
+`	type "effect"'
+`	no_pm "true"'
+`	event_flags	"15"' # trapping PRE/POST_PMU/PMD events
+`	event_type	"1"' # 1 for DAPM event for detect component
```
Comment (ranj063) [inline tools/topology/m4/detect.m4]: @keyonjie don't we need stream_name here?

File: tools/topology/sof/pipe-kfbm-capture.m4 [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+indir(`define', concat(`PIPELINE_SOURCE_', PIPELINE_ID), N_KPBM(0, PIPELINE_ID))
+indir(`define', concat(`PIPELINE_PCM_', PIPELINE_ID), N_PCMC(PCM_ID))
+
+#
+# PCM Configuration
+#
+dnl PCM_CAPTURE_ADD(name, pipeline, capture)
+PCM_CAPTURE_ADD(DMIC01, PIPELINE_ID, KPBM Capture PCM_ID)
+
+PCM_CAPABILITIES(KPBM Capture PCM_ID, `S32_LE,S24_LE,S16_LE', 16000, 16000, 2, PIPELINE_CHANNELS, 2, 16, 192, 16384, 65536, 65536)
```
Comment (ranj063) [inline tools/topology/sof/pipe-kfbm-capture.m4]: @keyonjie @singalsu @plbossart I wondering if we should restrict the PCM capabilities to support only the S16_LE here?

File: tools/topology/sof-apl-keyword-detect.m4 [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+
+dnl PIPELINE_PCM_ADD(pipeline,
+dnl     pipe id, pcm, max channels, format,
+dnl     frames, deadline, priority, core)
+
+# Passthrough capture pipeline 1 on PCM 0 using max 2 channels.
+# Schedule 320 frames per 1000us deadline on core 0 with priority 0
+PIPELINE_PCM_ADD(sof/pipe-kfbm-capture.m4,
+	1, 0, 2, s32le,
+	320, 1000, 0, 0)
```
Comment (ranj063) [inline tools/topology/sof-apl-keyword-detect.m4]: @keyonjie 20000 for the deadline? This is why we need the macro :)

File: tools/topology/sof-apl-keyword-detect.m4 [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+dnl     buffer, periods, format,
+dnl     frames, deadline, priority, core)
+
+
+# capture DAI is DMIC 0 using 2 periods
+# Buffers use s16le format, with 320 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-capture.m4,
+	1, DMIC, 0, NoCodec-6,
+	PIPELINE_SINK_1, 2, s16le,
+	320, 1000, 0, 0)
```
Comment (ranj063) [inline tools/topology/sof-apl-keyword-detect.m4]: same here for the deadline

File: tools/topology/sof-apl-keyword-detect.m4 [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+                dapm(PIPELINE_SINK_2, PIPELINE_SOURCE_1)
+		dapm(PIPELINE_PCM_1, PIPELINE_DETECT_2)
+        ]
+}
+
+#
+# BE configurations - overrides config in ACPI if present
+#
+
+#TODO: Change to DMIC 1
```
Comment (ranj063) [inline tools/topology/sof-apl-keyword-detect.m4]: @keyonjie @singalsu is this comment still valid? should it be DMIC 1 or DMIC 0

File: tools/topology/m4/detect.m4 (hunk 16 above)
Comment (keyonjie) [reply tools/topology/m4/detect.m4]: the get stream worked without it, I can add it if it is needed in driver.

File: tools/topology/sof/pipe-kfbm-capture.m4 (hunk 17 above)
Comment (keyonjie) [reply tools/topology/sof/pipe-kfbm-capture.m4]: let me limit it to S16_LE only.

File: tools/topology/sof-apl-keyword-detect.m4 (hunk 18 above)
Comment (keyonjie) [reply tools/topology/sof-apl-keyword-detect.m4]: I think the deadline is not the period team, it is for scheduler, so no need to change deadline accordingly?

Comment (ranj063) [reply tools/topology/sof-apl-keyword-detect.m4]: @keyonjie the deadline should change too to maintain 16KHz isnt it?

File: tools/topology/m4/detect.m4 [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+`	tuples."word" {'
+`		SOF_TKN_COMP_PERIOD_SINK_COUNT'		STR($3)
+`		SOF_TKN_COMP_PERIOD_SOURCE_COUNT'	STR($4)
+`	}'
+`}'
+`SectionVendorTuples."'N_DETECT($1)`_detect_process_tuples_str" {'
+`	tokens "sof_process_tokens"'
+`	tuples."string" {'
+`		SOF_TKN_PROCESS_TYPE'	STR(concat($5, _DETECT))
+`		SOF_TKN_PROCESS_SUBTYPE	"SIGSINK"'
```
Comment (ranj063) [inline tools/topology/m4/detect.m4]: @keyonjie remove subtype here?

File: tools/topology/sof-apl-keyword-detect.m4 (hunk 18 above)
Comment (keyonjie) [reply tools/topology/sof-apl-keyword-detect.m4]: @tlauda can you confirm that if we need change deadline for different period size?

File: tools/topology/m4/detect.m4 (hunk 21 above)
Comment (keyonjie) [reply tools/topology/m4/detect.m4]: let me remove it.

File: tools/topology/sof-apl-keyword-detect.m4 (hunk 18 above)
Comment (ranj063) [reply tools/topology/sof-apl-keyword-detect.m4]: @tlauda @keyonjie this PR is ready for merging after confirming if the deadline needs to change or not.

Comment (tlauda) [reply tools/topology/sof-apl-keyword-detect.m4]: @ranj063 @keyonjie Pipeline will be scheduled every 20 ms, so the deadline should be also increased to 20000. 1000 will also work, but in case of many pipelines it's better not to risk it.

Comment (keyonjie) [reply tools/topology/sof-apl-keyword-detect.m4]: @tlauda if so, we actually don't need this param passed in to FW, we can set it to period_time inside FW directlly.

Comment (keyonjie) [reply tools/topology/sof-apl-keyword-detect.m4]: anyway, let me change to use macros for it ATM.

**Comment by keyonjie** [comment]
updated. @tlauda @ranj063 @lgirdwood @lbetlej I think it's good to be merged now.

**Comment by keyonjie** [comment]
@lgirdwood @mmaka1 ping for reviewing and merging then we can focus on the integration and component feature enabling/debugging with same base.
